### PR TITLE
Add gateway address to the returned IPFS api

### PR DESCRIPTION
**Pull request by haadcode**
This PR will add a properties to get the address the gateway was started at and can be accessed with:

```javascript
ipfs.gatewayHost
ipfs.gatewayPort
```

### CODE REVIEW

**Comment by dignifiedquire** [comment]
CI is very unhappy here

**Comment by haadcode** [comment]
Fixed!

File: src/node.js [hunk 1]
```diff
@@ -198,6 +198,7 @@ class Node {
         },
         data: (data) => {
           const match = String(data).trim().match(/API server listening on (.*)/)
+          const gwmatch = String(data).trim().match(/Gateway (.*) listening on (.*)/)
```
Comment (dignifiedquire) [inline src/node.js]: Lets not duplicate work. The trim and string cast are happening twice now.

File: src/node.js [hunk 2]
```diff
@@ -206,6 +207,13 @@ class Node {
             this.api.apiHost = addr.address
             this.api.apiPort = addr.port
 
+            if (gwmatch) {
+              this.gatewayAddr = gwmatch[2]
+              const addr = multiaddr(this.gatewayAddr).nodeAddress()
```
Comment (dignifiedquire) [inline src/node.js]: How about only exposing `this.gatewayAddr = multiaddr(gwmatch[2])`. That way we have the standard multiaddr we are using everywhere and you can pull the details you need at any point.

File: src/node.js (hunk 1 above)
Comment (haadcode) [inline src/node.js]: Will fix this.

File: src/node.js (hunk 2 above)
Comment (haadcode) [inline src/node.js]: I believe following the current convention as per api host and port is a better approach for consistency.

We can do what you propose in separate PR, bearing in mind it would be a public API change.

Comment (dignifiedquire) [inline src/node.js]: how is the a breaking change, currently the `gatewayAddr` doesn't exist from what I can tell?

Comment (haadcode) [inline src/node.js]: Jumped some steps here :) Let me try again:

We currently have (with this PR):
```javascript
ipfs.apiHost
ipfs.apiPort
ipfs.gatewayHost
ipfs.gatewayPort
```

If we would expose the addresses as multiaddr, we should make sure the properties are consistent:
```javascript
ipfs.apiAddr
ipfs.gatewayAddr
```

So with that, I would get rid of `ipfs.apiHost` and `ipfs.apiPort`, and that would be the breaking change.

Does that make sense? We could also add both `ipfs.gatewayAddr` and `ipfs.apiAddr` in this PR, but then we'd have the `apiHost` and `apiPort` still there and I think that'd be slightly more confusion to have same thing as multiple properties.

Let me know what you think!

Comment (dignifiedquire) [inline src/node.js]: How about exposing `gatewayAddr` and `apiAddr` as the main thing, and making the others simple helper functions which just depend on those?

Comment (haadcode) [inline src/node.js]: Changed the code now to expose `apiAddr` and `gatewayAddr` in the `Node` and to keep api compatibility, keep the separate variables for host and port in the `ipfs` instance. I'm not a huge fan of the monkey-patching solution here as it changed the public API of js-ipfs-api, but that's how it's always been, so I don't want to break that now.

@dignifiedquire take a look at let me know if you think this is good to go.

**Comment by haadcode** [comment]
As noticed on IRC, this will change the public property `apiAddr` from string to multiaddr. We should prolly bump the version? Should I bump it to 0.19.0?

**Comment by dignifiedquire** [comment]
version bump will happen on release.

**Comment by diasdavid** [comment]
- [ ] Needs deps to be updated (npm-go-ipfs)

Is there PR to cope with the breaking change proposed here?

**Comment by haadcode** [comment]
> Are there PRs to cope with the breaking change proposed here?

This will change the public API of `ipfsd-ctl`, not the ipfs API returned by it. Ie. nothing should break for js-ipfs-api. Do we test this module somewhere else than here in its own tests?

**Comment by diasdavid** [comment]
Just to be sure, can you check if js-ipfs-api runs smoothly with these changes?

**Comment by haadcode** [comment]
> Just to be sure, can you check if js-ipfs-api runs smoothly with these changes?

Sure thing! I'll add tests to js-ipfs-api for this today.

**Comment by diasdavid** [comment]
https://github.com/ipfs/js-ipfs-api/pull/524/files#r101501630
